File: product-service/src/main/java/com/intexsoft/webshop/productservice/model/Product.java
package com.intexsoft.webshop.productservice.model;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.FieldDefaults;

import java.util.LinkedHashSet;
import java.util.Set;

@Entity
@Table(name = "product")
@Data
@NoArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    Long id;
    @Column(name = "name", nullable = false)
    String name;
    @ManyToOne(fetch = FetchType.LAZY)
    Subcategory subcategory;
    @ManyToOne(fetch = FetchType.LAZY)
    Vendor vendor;
    @OneToMany(
            mappedBy = "product",
            fetch = FetchType.LAZY,
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    Set<Image> images = new LinkedHashSet<>();
    @OneToMany(
            mappedBy = "product",
            fetch = FetchType.LAZY,
            cascade = CascadeType.ALL
    )
    Set<AttributeValue> attributeValues = new LinkedHashSet<>();

    public void addImage(Image image) {
        images.add(image);
        image.setProduct(this);
    }

    public void removeImage(Image image) {
        images.remove(image);
        image.setProduct(null);
    }

    public void addAttributeValue(AttributeValue attributeValue) {
        attributeValues.add(attributeValue);
        attributeValue.setProduct(this);
    }

    public void removeAttributeValue(AttributeValue attributeValue) {
        attributeValues.remove(attributeValue);
        attributeValue.setProduct(null);
    }
}

File: product-service-kt/src/main/kotlin/com/intexsoft/webshop/productservicekt/model/Product.kt
package com.intexsoft.webshop.productservicekt.model

import jakarta.persistence.*

@Entity
@Table(name = "product")
class Product(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0,
    @Column(name = "name", nullable = false)
    var name: String,
    @Column(name = "orderQuantity")
    var orderQuantity: Short = 0,
    @ManyToOne(fetch = FetchType.LAZY)
    var subcategory: Subcategory,
    @ManyToOne(fetch = FetchType.LAZY)
    var vendor: Vendor,
    @OneToMany(mappedBy = "product", fetch = FetchType.LAZY, cascade = [CascadeType.ALL], orphanRemoval = true)
    var images: MutableList<Image>?,
    @OneToMany(mappedBy = "product", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    var attributeValues: MutableList<AttributeValue>?
) {
    fun addImage(image: Image) {
        if (images == null) images = mutableListOf()
        images!!.add(image)
        image.product = this
    }

    fun removeImage(image: Image) {
        images?.remove(image)
        image.product = null
    }

    fun addAttributeValue(attributeValue: AttributeValue) {
        if (attributeValues == null) attributeValues = mutableListOf()
        attributeValues!!.add(attributeValue)
        attributeValue.product = this
    }

    fun removeAttributeValue(attributeValue: AttributeValue) {
        attributeValues?.remove(attributeValue)
        attributeValue.product = null
    }
}